File: main/kth-largest-element-in-a-stream-lc/kth-largest-element-in-a-stream-lc.py
class KthLargest:
    """Data strcure to finds kth largest element in a stream, for fixed k."""
    
    __slots__ = ('_k', '_heap')
    
    def __init__(self, k: int, nums: List[int]):
        """Creates the data structure with the given k, starting with nums."""
        self._k = k
        self._heap = nums[:self._k]
        
        heapq.heapify(self._heap)
        
        for val in nums[self._k:]:
            heapq.heappushpop(self._heap, val)
        
        assert len(self._heap) <= self._k

    def add(self, val: int) -> int:
        """Receives val from the stream. Finds the new kth largest element."""
        if len(self._heap) == self._k:
            heapq.heappushpop(self._heap, val)
        else:
            assert len(self._heap) < self._k, 'kept too many elements'
            heapq.heappush(self._heap, val)
        
        return self._heap[0]
        

# Your KthLargest object will be instantiated and called as such:
# obj = KthLargest(k, nums)
# param_1 = obj.add(val)

File: main/kth-largest-element-in-a-stream-lc/kth-largest-element-in-a-stream-lc.cpp
class KthLargest {
public:
    KthLargest(int k, const vector<int>& nums);
    
    int add(int val);

private:
    std::priority_queue<int, std::vector<int>, std::greater<>> heap_;
    const int k_;
};

KthLargest::KthLargest(const int k, const vector<int>& nums)
    : heap_(cbegin(nums), (nums.size() <= k ? cend(nums) : cbegin(nums) + k)),
      k_{k}
{
    if (nums.size() <= k) return;
    
    for_each(cbegin(nums) + k, cend(nums), [this](const int val) noexcept {
        heap_.push(val);
        heap_.pop();
    });
}

int KthLargest::add(const int val)
{
    heap_.push(val);
    
    if (heap_.size() == k_ + 1) {
        heap_.pop();
    } else {
        assert(heap_.size() <= k_);
    }
    
    return heap_.top();
}

/**
 * Your KthLargest object will be instantiated and called as such:
 * KthLargest* obj = new KthLargest(k, nums);
 * int param_1 = obj->add(val);
 */
